### PR TITLE
Fix/178/for loop parameters

### PR DESCRIPTION
**Pull request by msangel**
Update the `For` tag implementation to the full standard (I hope so). 
important changes:

- `continue` now residents in the registry, so no interference with variables.  
- `forloop.parentloop` is now supported
- `reversed` option is now supported
- exceptions from the loop over array body are no longer ignored silently 

and minor edge cases support from the liquid test suite

### CODE REVIEW

File: src/main/java/liqp/LValue.java [hunk 1]
```diff
@@ -174,7 +176,19 @@ public boolean asBoolean(Object value) {
...
     public Number asNumber(Object value) throws NumberFormatException {
 
-        if(value instanceof Number) {
+        if (value == null) {
+            return 0;
+        }
+
+        if (value instanceof String) {
+            Matcher matcher = Pattern.compile("[\\d.]+").matcher((String) value);
+            if (matcher.find()) {
```
Comment (bkiers) [inline src/main/java/liqp/LValue.java:185]: Do you want to do `find()` here? In that case, when the input is `"1.2foo"`, `matcher.find()` will return `true` and `1.2` will be returned. When you use `matcher.matches()`, it will return `false`the input `"1.2foo"`. However, when using `matcher.matches()`, `return 0;` will be performed. So then `"1.2foo"` will become `0`. Is that also how Ruby's Liquid engine handles it?

File: src/main/java/liqp/tags/For.java [hunk 2]
```diff
@@ -83,188 +103,239 @@ private Object renderArray(String id, TemplateContext context, LNode... tokens)
...
 
-    private Object renderRange(String id, TemplateContext context, LNode... tokens) {
+    private Object renderRange(String id, TemplateContext context, String tagName, boolean reversed, LNode... tokens) {
 
         StringBuilder builder = new StringBuilder();
 
-        // attributes start from index 5
-        Map<String, Integer> attributes = getAttributes(5, context, tokens);
+        // attributes start from index 6
+        Map<String, Integer> attributes = getAttributes(7, context, tagName, tokens);
```
Comment (bkiers) [inline src/main/java/liqp/tags/For.java:212]: The comment `attributes start from index 6` does not match the code: `getAttributes(7, ...)`

**Review by bkiers — APPROVED** [review]
Nice work! Grammar changes look good. Just 2 small comments.

File: src/main/java/liqp/LValue.java (hunk 1 above)
Comment (msangel) [reply src/main/java/liqp/LValue.java:185]: This is a very good question. 
In short, there are some rules around that in Ruby. Here's a short article; https://code-maven.com/convert-string-to-number-in-ruby . I tried to mimic that as close as possible and fail, so decided to move this out of this task scope. The current implementation is a trade-off - it covers the most popular case. 
Full list of cases for conversion to int can be found here: https://github.com/jruby/jruby/blob/master/spec/ruby/core/string/to_i_spec.rb
Regarding `1.2foo` , in ruby itself: 
```ruby
"1.2foo".to_i # => 1
"1.2foo".to_f # => 1.2
# BUT
"5e4".to_i  # => 5
"5e4".to_f # => 50000.0 (accidentally the java double parses recognize this)
```
So this code does the same for simple cases.
In the liquid engine, it's all is more around how the tags work with the value. And the good news is: both we and liquid in most cases do convert this "1.2foo" to `1`.  
Reimplementing ruby`s string to int and string to float is too much for this I believe, as this is java library in the end. 
There is no simple answer on this problem.

Comment (msangel) [reply src/main/java/liqp/LValue.java]: Ok. I see this is a problematic place in many aspects, so simply removed this and left previous implementation.

File: src/main/java/liqp/tags/For.java (hunk 2 above)
Comment (msangel) [reply src/main/java/liqp/tags/For.java:212]: Fixed.

File: src/main/java/liqp/LValue.java (hunk 1 above)
Comment (msangel) [reply src/main/java/liqp/LValue.java]: This problem can be discussed in the scope of a new issue if will be really actual.

**Comment by msangel** [comment]
Closes https://github.com/bkiers/Liqp/issues/178
